### DISCUSSION TITLE
how to load the .stl files?

### DISCUSSION THREAD

**Discussion by jqdsn** [libigl/libigl#2116] [Q&A] [answered]
I want to load the Thingi10k dataset. But the readSTL.h files can't to load by filename of .stl files, there is only two kinds of functions.
 `IGL_INLINE bool readSTL(
    std::istream &input,
    Eigen::PlainObjectBase<DerivedV> & V,
    Eigen::PlainObjectBase<DerivedF> & F,
    Eigen::PlainObjectBase<DerivedN> & N);

 template <typename DerivedV, typename DerivedF, typename DerivedN>
  IGL_INLINE bool readSTL(
    FILE * fp,
    Eigen::PlainObjectBase<DerivedV> & V,
    Eigen::PlainObjectBase<DerivedF> & F,
    Eigen::PlainObjectBase<DerivedN> & N);
}`

I try to use the  two kinds of functions,but it is unsuccessful.
my code:

` FILE *fp1 = fopen(TUTORIAL_SHARED_PATH "/34785.obj", "r");
  FILE *fp2 = fopen(TUTORIAL_SHARED_PATH "/34783.obj", "r");
  igl::readSTL(fp1, VA, FA, NA);
  igl::readSTL(fp2, VB, FB, NB);`

error:
`igl.lib(read_triangle_mesh.obj) : error LNK2019: 无法解析的外部符号 "bool __cdecl igl::readSTL<class Eigen::Matrix`

**Comment by alecjacobson** [comment]
huh. @jdumas do you know what's up? I think the string input header disappeared in your change https://github.com/libigl/libigl/commit/1e905cba3d5e7d82487a6826d0d3fa7cd47de114#diff-b52f1e6058e4af83e1d76954459e8247f46641e1d9cd0de098bef9f8528be024

does a path string get somehow converted to a istream?

**Reply by jdumas** [reply]
Hmm I think the goal was to have a single function signature that can load both from memory and from a file. Typically a function that operates on a `std::istream` can do that. The only reason I've added the `FILE *` overload was because of the `read_triangle_mesh()` function.

To load a STL from a file you can simply do:
```c++
{
  std::ifstream in("my_mesh.stl");
  igl::readSTL(in, V, F, N);
}
```

I suppose you _could_ add an overload that takes a `std::string` or a `std::filesystem::path` as input, but that'd be 2x the number of function signatures to expose for little added benefits.

**Reply by jqdsn** [reply]
I use the method,but it's unsuccessful.
my code:
`std::ifstream in1(TUTORIAL_SHARED_PATH "/34785.obj");
  std::ifstream in2(TUTORIAL_SHARED_PATH "/34783.obj");
  
  igl::readSTL(in1, VA, FA, NA);
  igl::readSTL(in2, VB, FB, NB);`

error:
 `main.obj : error  LNK2019	无法解析的外部符号 "bool __cdecl igl::readSTL<class Eigen::Matrix<double,-1,-1,0,-1,-1>,class Eigen::Matrix<int,-1,-1,0,-1,-1>,class Eigen::Matrix<double,-1,-1,0,-1,-1> >(class std::basic_istream<char,struct std::char_traits<char> > &,class Eigen::PlainObjectBase<class Eigen::Matrix<double,-1,-1,0,-1,-1> > &,class Eigen::PlainObjectBase<class Eigen::Matrix<int,-1,-1,0,-1,-1> > &,class Eigen::PlainObjectBase<class Eigen::Matrix<double,-1,-1,0,-1,-1> > &)" (??$readSTL@V?$Matrix@N$0?0$0?0$0A@$0?0$0?0@Eigen@@V?$Matrix@H$0?0$0?0$0A@$0?0$0?0@2@V12@@igl@@YA_NAEAV?$basic_istream@DU?$char_traits@D@std@@@std@@AEAV?$PlainObjectBase@V?$Matrix@N$0?0$0?0$0A@$0?0$0?0@Eigen@@@Eigen@@AEAV?$PlainObjectBase@V?$Matrix@H$0?0$0?0$0A@$0?0$0?0@Eigen@@@4@1@Z)，函数 main 中引用了该符号	609_Boolean	G:\bool operator experiment\libigl-main\build\tutorial\main.obj	1	
`

**Reply by jdumas** [reply]
Ah it looks like we don't have any explicit template instantiation for the `std::istream` version of the function... You could try with the header-only version of libigl for now.

**Comment by alecjacobson** [comment, marked as answer]
You can also use read_triangle_mesh with a path to the STL file.

**Reply by jqdsn** [reply]
I use read_triangle_mesh with a path to the STL file. it's ok,
thank you very much.
